apis/temperature/src/lib.rs

```diff
@@ -0,0 +1,81 @@
+#![no_std]
+
+use core::cell::Cell;
+use libtock_platform::{
+    share, subscribe::OneId, DefaultConfig, ErrorCode, Subscribe, Syscalls, Upcall,
+};
+
+pub struct Temperature<S: Syscalls>(S);
+
+impl<S: Syscalls> Temperature<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// Initiate a temperature measurement.
+    ///
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_temperature() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_TEMP, 0, 0).to_result()
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share, F: Fn(i32)>(
+        listener: &'share TemperatureListener<F>,
+        subscribe: share::Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Initiate a synchronous temperature measurement.
+    /// Returns Ok(temperature_value) if the operation was successful
+    /// temperature_value is returned in hundreds of centigrades
+    pub fn read_temperature_sync() -> Result<i32, ErrorCode> {
+        let temperature_cell: Cell<Option<i32>> = Cell::new(None);
+        let listener = TemperatureListener(|temp_val| {
+            temperature_cell.set(Some(temp_val));
+        });
```

Comment:
If you do the following:

1. Change `register_listener` to take a `F: Upcall<OneId<DRIVER_NUM, 0>>`
2. Change `temperature_cell` to be a `Cell<Option<(i32,)>>`

then you should be able to pass `&temperature_cell` directly to `register_listener` and remove `TemperatureListener` entirely.

Reply:
I tried that, but the `Upcall` trait is not implemented for `Cell<Option<(i32,)>>`.It is implemented for `Cell<Option<(u32,)>>` though, so should I use it and cast the value to i32 afterwards?

Reply:
I would not go this way, as the asynchronous call will then return a `u32` instead of `i32`.

Reply:
> I tried that, but the `Upcall` trait is not implemented for `Cell<Option<(i32,)>>`.It is implemented for `Cell<Option<(u32,)>>` though, so should I use it and cast the value to i32 afterwards?

Yes.

Reply:
@jrvanwhy I might be wrong, but this will not allows us to make this conversion when using the asynchronous API.

Reply:
> @jrvanwhy I might be wrong, but this will not allows us to make this conversion when using the asynchronous API.

Oh, I see the issue now. Yes, that would break the asynchronous API. Lets leave this as-is.